File: app/src/main/java/com/scheffer/erik/videogamewishlist/activities/GameSearchListActivity.java
package com.scheffer.erik.videogamewishlist.activities;

import android.database.Cursor;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.scheffer.erik.videogamewishlist.BuildConfig;
import com.scheffer.erik.videogamewishlist.R;
import com.scheffer.erik.videogamewishlist.database.WishlistContract;
import com.scheffer.erik.videogamewishlist.models.Game;
import com.scheffer.erik.videogamewishlist.models.Genre;
import com.scheffer.erik.videogamewishlist.models.IGDBGame;
import com.scheffer.erik.videogamewishlist.models.Platform;
import com.scheffer.erik.videogamewishlist.models.Theme;
import com.scheffer.erik.videogamewishlist.recyclerviewadapters.GameRecyclerViewAdapter;

import org.json.JSONArray;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import callback.OnSuccessCallback;
import wrapper.Endpoints;
import wrapper.IGDBWrapper;
import wrapper.Parameters;
import wrapper.Version;

public class GameSearchListActivity extends AppCompatActivity {

    public static final String GAME_TITLE_EXTRA = "game-title";
    public static final String PLATFORM_EXTRA = "platform";
    public static final String GENRE_EXTRA = "genre";
    public static final String THEME_EXTRA = "theme";
    public static final String MINIMUM_RATING_EXTRA = "minimum-rating";
    public static final String MAXIMUM_RATING_EXTRA = "maximum-rating";

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.search_list_info_text)
    TextView infoText;
    @BindView(R.id.game_list)
    RecyclerView gamesRecyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_search_list);
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        gamesRecyclerView.setVisibility(View.GONE);
        infoText.setVisibility(View.VISIBLE);

        infoText.setText(R.string.loading_game_list);

        final IGDBWrapper wrapper = new IGDBWrapper(BuildConfig.API_KEY,
                                                    Version.STANDARD,
                                                    false);

        Parameters params = new Parameters()
                .addFields("*")
                .addOrder("rating:desc")
                .addLimit("50");

        Platform platform = getIntent().getParcelableExtra(PLATFORM_EXTRA);
        if (platform != null) {
            params.addFilter("[platforms][eq]=" + platform.getId());
        }

        Genre genre = getIntent().getParcelableExtra(GENRE_EXTRA);
        if (genre != null) {
            params.addFilter("[genres][eq]=" + genre.getId());
        }

        Theme theme = getIntent().getParcelableExtra(THEME_EXTRA);
        if (theme != null) {
            params.addFilter("[theme][eq]=" + theme.getId());
        }

        int minimumRating = getIntent().getIntExtra(MINIMUM_RATING_EXTRA, -1);
        if (minimumRating > 0) {
            params.addFilter("[rating][gte]=" + minimumRating);
        }

        int maximumRating = getIntent().getIntExtra(MAXIMUM_RATING_EXTRA, 101);
        if (maximumRating < 100) {
            params.addFilter("[rating][lte]=" + maximumRating);
        }

        String gameTitle = getIntent().getStringExtra(GAME_TITLE_EXTRA);
        if (!gameTitle.isEmpty()) {
            params.addSearch(gameTitle);
            wrapper.search(Endpoints.GAMES, params, new MySuccessCallback());
        } else {
            wrapper.games(params, new MySuccessCallback());
        }

    }

    private class MySuccessCallback implements OnSuccessCallback {

        @Override
        public void onSuccess(JSONArray result) {

            Type listType = new TypeToken<ArrayList<IGDBGame>>() {}.getType();
            ArrayList<IGDBGame> fromGson = new GsonBuilder()
                    .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                    .create().fromJson(result.toString(), listType);


            final ArrayList<Game> games = new ArrayList<>();
            for (IGDBGame igdbGame : fromGson) {
                Game game = new Game();
                game.setId(igdbGame.getId());
                game.setName(igdbGame.getName());
                game.setSummary(igdbGame.getSummary());
                game.setRating(igdbGame.getRating());
                game.setCover(igdbGame.getCover());
                game.setVideos(igdbGame.getVideos());

                List<Platform> platforms = new ArrayList<>();
                if (igdbGame.getPlatforms() != null) {
                    for (Long platformId : igdbGame.getPlatforms()) {
                        Cursor cursor = getContentResolver()
                                .query(WishlistContract.PlatformEntry.CONTENT_URI.buildUpon()
                                                                                 .appendPath(
                                                                                         String.valueOf(
                                                                                                 platformId))
                                                                                 .build(),
                                       null,
                                       null,
                                       null,
                                       null);
                        if (cursor != null) {
                            if (cursor.getCount() > 0) {
                                cursor.moveToFirst();
                                platforms.add(new Platform(platformId,
                                                           cursor.getString(cursor.getColumnIndex(
                                                                   WishlistContract.PlatformEntry.COLUMN_NAME))));
                            }
                            cursor.close();
                        }
                    }
                }
                game.setPlatforms(platforms);

                List<Genre> genres = new ArrayList<>();
                if (igdbGame.getGenres() != null) {
                    for (Long genreId : igdbGame.getGenres()) {
                        Cursor cursor = getContentResolver()
                                .query(WishlistContract.GenreEntry.CONTENT_URI.buildUpon()
                                                                              .appendPath(
                                                                                      String.valueOf(
                                                                                              genreId))
                                                                              .build(),
                                       null,
                                       null,
                                       null,
                                       null);
                        if (cursor != null) {
                            if (cursor.getCount() > 0) {
                                cursor.moveToFirst();
                                genres.add(new Genre(genreId,
                                                     cursor.getString(cursor.getColumnIndex(
                                                             WishlistContract.ThemeEntry.COLUMN_NAME))));
                            }
                            cursor.close();
                        }
                    }
                }
                game.setGenres(genres);

                List<Theme> themes = new ArrayList<>();
                if (igdbGame.getThemes() != null) {
                    for (Long themeId : igdbGame.getThemes()) {
                        Cursor cursor = getContentResolver()
                                .query(WishlistContract.ThemeEntry.CONTENT_URI.buildUpon()
                                                                              .appendPath(
                                                                                      String.valueOf(
                                                                                              themeId))
                                                                              .build(),
                                       null,
                                       null,
                                       null,
                                       null);
                        if (cursor != null) {
                            if (cursor.getCount() > 0) {
                                cursor.moveToFirst();
                                themes.add(new Theme(themeId,
                                                     cursor.getString(cursor.getColumnIndex(
                                                             WishlistContract.ThemeEntry.COLUMN_NAME))));
                            }
                            cursor.close();
                        }
                    }
                }
                game.setThemes(themes);

                games.add(game);
            }

            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    if (games.isEmpty()) {
                        infoText.setText(R.string.no_games_found);
                    } else {
                        gamesRecyclerView.setVisibility(View.VISIBLE);
                        infoText.setVisibility(View.GONE);

                        gamesRecyclerView.setLayoutManager(new LinearLayoutManager(
                                GameSearchListActivity.this));
                        gamesRecyclerView.setAdapter(new GameRecyclerViewAdapter(games));
                    }
                }
            });
        }

        @Override
        public void onError(Exception error) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    infoText.setText(R.string.error_games_from_server);
                }
            });
            error.printStackTrace();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/scheffer/erik/videogamewishlist/activities/GameSearchListActivity.kt
package com.scheffer.erik.videogamewishlist.activities

import APICalypse
import IGDBWrapper
import RequestException
import Sort
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import com.scheffer.erik.videogamewishlist.BuildConfig
import com.scheffer.erik.videogamewishlist.R
import com.scheffer.erik.videogamewishlist.converters.fromProtoGameToGame
import com.scheffer.erik.videogamewishlist.models.Game
import com.scheffer.erik.videogamewishlist.models.Genre
import com.scheffer.erik.videogamewishlist.models.Platform
import com.scheffer.erik.videogamewishlist.models.Theme
import com.scheffer.erik.videogamewishlist.recyclerviewadapters.GameRecyclerViewAdapter
import games
import kotlinx.android.synthetic.main.activity_game_search_list.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class GameSearchListActivity : AppCompatActivity() {

    var games: MutableList<Game> = ArrayList()
    var gameRecyclerViewAdapter: GameRecyclerViewAdapter? = null

    private var reverseOrder: Boolean = false
    private val layoutManager = LinearLayoutManager(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game_search_list)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        game_list.visibility = View.GONE
        search_list_info_text.visibility = View.VISIBLE

        search_list_info_text.setText(R.string.loading_game_list)

        savedInstanceState?.let {
            layoutManager.onRestoreInstanceState(it.getParcelable(LIST_STATE_KEY))
            @Suppress("UNCHECKED_CAST")
            games = it.getParcelableArray(GAME_LIST_KEY).toMutableList() as MutableList<Game>
        }

        if (games.isEmpty()) {
            loadGames()
        } else {
            initializeRecyclerView()
        }
    }

    private fun loadGames() {
        IGDBWrapper.userkey = BuildConfig.API_KEY
        var params = APICalypse()
                .fields("*, cover.*, videos.*")
                .limit(50)
                .sort("rating", Sort.DESCENDING)

        val whereParams = mutableListOf<String>()

        intent.getParcelableExtra<Platform>(PLATFORM_EXTRA)?.let { platform ->
            whereParams.add("platforms = ${platform.id}")
        }

        intent.getParcelableExtra<Genre>(GENRE_EXTRA)?.let { genre ->
            whereParams.add("genres = ${genre.id}")
        }

        intent.getParcelableExtra<Theme>(THEME_EXTRA)?.let { theme ->
            whereParams.add("themes = ${theme.id}")
        }

        val minimumRating = intent.getIntExtra(MINIMUM_RATING_EXTRA, -1)
        if (minimumRating > 0) {
            whereParams.add("rating >= $minimumRating")
        }

        val maximumRating = intent.getIntExtra(MAXIMUM_RATING_EXTRA, 101)
        if (maximumRating < 100) {
            whereParams.add("rating <= $maximumRating")
        }

        val gameTitle = intent.getStringExtra(GAME_TITLE_EXTRA) ?: ""
        if (gameTitle.isNotEmpty()) {
            whereParams.add("name ~ *\"$gameTitle\"*")
        }

        if (whereParams.isNotEmpty()) {
            params = params.where(whereParams.joinToString(" & ") + " & rating != null")
        } else {
            params = params.where("rating != null")
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                games = IGDBWrapper.games(params).map(::fromProtoGameToGame).toMutableList()
                runOnUiThread {
                    if (games.isEmpty()) {
                        search_list_info_text.setText(R.string.no_games_found)
                    } else {
                        initializeRecyclerView()
                    }
                }
            } catch (error: RequestException) {
                runOnUiThread { search_list_info_text.setText(R.string.error_games_from_server) }
                error.printStackTrace()
            }
        }
    }

    private fun initializeRecyclerView() {
        game_list.visibility = View.VISIBLE
        search_list_info_text.visibility = View.GONE

        game_list.layoutManager = layoutManager
        gameRecyclerViewAdapter = GameRecyclerViewAdapter(games)
        game_list.adapter = gameRecyclerViewAdapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_sort_options, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
            R.id.sort_name -> {
                if (reverseOrder) games.sortByDescending { it.name }
                else games.sortBy { it.name }

                gameRecyclerViewAdapter?.notifyDataSetChanged()
                reverseOrder = !reverseOrder
                return true
            }
            R.id.sort_rating -> {
                if (reverseOrder) games.sortByDescending { it.rating }
                else games.sortBy { it.rating }

                gameRecyclerViewAdapter?.notifyDataSetChanged()
                reverseOrder = !reverseOrder
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(state: Bundle) {
        super.onSaveInstanceState(state)
        state.putParcelable(LIST_STATE_KEY, layoutManager.onSaveInstanceState())
        state.putParcelableArray(GAME_LIST_KEY, games.toTypedArray())
    }

    companion object {
        const val GAME_TITLE_EXTRA = "game-title"
        const val PLATFORM_EXTRA = "platform"
        const val GENRE_EXTRA = "genre"
        const val THEME_EXTRA = "theme"
        const val MINIMUM_RATING_EXTRA = "minimum-rating"
        const val MAXIMUM_RATING_EXTRA = "maximum-rating"

        private const val LIST_STATE_KEY = "list-state"
        private const val GAME_LIST_KEY = "game-list"
    }
}
